staging/src/k8s.io/mount-utils/resizefs_linux.go

```diff
@@ -84,3 +86,134 @@ func (resizefs *ResizeFs) xfsResize(deviceMountPath string) (bool, error) {
+	}
+	return true, nil
+}
+func (resizefs *ResizeFs) getDeviceSize(devicePath string) (uint64, error) {
+	output, err := resizefs.exec.Command("blockdev", "--getsize64", devicePath).CombinedOutput()
+	outStr := strings.TrimSpace(string(output))
+	if err != nil {
+		return 0, fmt.Errorf("failed to read size of device %s: %s: %s", devicePath, err, outStr)
+	}
+	size, err := strconv.ParseUint(outStr, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse size of device %s %s: %s", devicePath, outStr, err)
+	}
+	return size, nil
+}
+
+func (resizefs *ResizeFs) getExtSize(devicePath string) (uint64, uint64, error) {
+	output, err := resizefs.exec.Command("dumpe2fs", "-h", devicePath).CombinedOutput()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to read size of filesystem on %s: %s: %s", devicePath, err, string(output))
+	}
+
+	lines := strings.Split(string(output), "\n")
+	var blockSize, blockCount uint64
+
+	for _, line := range lines {
+		tokens := strings.Split(line, ":")
+		if len(tokens) != 2 {
+			continue
+		}
+		key, value := strings.TrimSpace(tokens[0]), strings.TrimSpace(tokens[1])
+		if key == "Block count" {
+			blockCount, err = strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return 0, 0, fmt.Errorf("failed to parse block count %s: %s", value, err)
+			}
+		}
+		if key == "Block size" {
+			blockSize, err = strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return 0, 0, fmt.Errorf("failed to parse block size %s: %s", value, err)
+			}
+		}
+	}
+	if blockSize == 0 {
+		return 0, 0, fmt.Errorf("could not find block size of device %s", devicePath)
+	}
+	if blockCount == 0 {
+		return 0, 0, fmt.Errorf("could not find block count of device %s", devicePath)
+	}
+	return blockSize, blockSize * blockCount, nil
+}
+
+func (resizefs *ResizeFs) getXFSSize(devicePath string) (uint64, uint64, error) {
+	output, err := resizefs.exec.Command("xfs_info", devicePath).CombinedOutput()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to read size of filesystem on %s: %s: %s", devicePath, err, string(output))
+	}
+
+	lines := strings.Split(string(output), "\n")
+	var blockSize, blockCount uint64
+
+	for _, line := range lines {
+
+		if !strings.HasPrefix(line, "data") {
+			continue
+		}
+		line = strings.ReplaceAll(line, ",", "")
+		tokens := strings.FieldsFunc(line, func(c rune) bool {
+			return c == ' ' || c == '='
+		})
+		for index := range tokens {
+			if strings.TrimSpace(tokens[index]) == "bsize" {
+				index = index + 1
+				blockSize, err = strconv.ParseUint(strings.TrimSpace(tokens[index]), 10, 64)
+				if err != nil {
+					return 0, 0, fmt.Errorf("failed to parse block size %s: %s", tokens[index], err)
+				}
+			}
+			if strings.TrimSpace(tokens[index]) == "blocks" {
+				index = index + 1
+				blockCount, err = strconv.ParseUint(strings.TrimSpace(tokens[index]), 10, 64)
+				if err != nil {
+					return 0, 0, fmt.Errorf("failed to parse block count %s: %s", tokens[index], err)
+				}
+			}
+		}
```

Comment:
same as before. We need unit tests for the parsing logic.

Reply:
updated